src/chunk.rs

```diff
@@ -16,36 +16,244 @@ impl From<Vec3> for ChunkDimensions {
+    Dense,
+    /// Specifies to construct a sparse sprite layer.
+    Sparse,
+}
+
+#[derive(Clone, Debug)]
+enum LayerKindInner {
+    Dense(DenseLayer),
+    Sparse(SparseLayer),
+}
+
+impl AsRef<dyn Layer> for LayerKindInner {
+    fn as_ref(&self) -> &dyn Layer {
+        match self {
+            LayerKindInner::Dense(s) => s,
+            LayerKindInner::Sparse(s) => s,
+        }
+    }
+}
+
+impl AsMut<dyn Layer> for LayerKindInner {
+    fn as_mut(&mut self) -> &mut dyn Layer {
+        match self {
+            LayerKindInner::Dense(s) => s,
+            LayerKindInner::Sparse(s) => s,
+        }
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct SpriteLayer {
+    inner: LayerKindInner,
+    entity: Option<Entity>,
+}
+
+#[derive(Debug, TypeUuid)]
+#[uuid = "47691827-0b89-4474-a14e-f2ea3c88320f"]
+#[doc(hidden)]
+pub struct Chunk {
+    coord: Vec2,
+    sprite_layers: Vec<Option<SpriteLayer>>,
+    // flags: Vec<u32>, // TODO
+    dimensions: Vec3,
+}
+
+impl Chunk {
+    pub(crate) fn new(coord: Vec2, dimensions: Vec3, max_layers: usize) -> Chunk {
+        Chunk {
+            coord,
+            sprite_layers: vec![None; max_layers],
+            dimensions,
+        }
+    }
+
+    pub(crate) fn coord(&self) -> Vec2 {
+        self.coord
+    }
+
+    pub(crate) fn max_size(&self) -> usize {
+        (self.dimensions.x() * self.dimensions.y() * self.dimensions.z()) as usize
+    }
+
+    pub(crate) fn layer_size(&self) -> usize {
+        (self.dimensions.x() * self.dimensions.y()) as usize
+    }
+
+    pub(crate) fn add_layer(&mut self, kind: LayerKind, z: usize) {
+        match kind {
+            LayerKind::Dense => {
+                let tiles = vec![Tile::new(0); self.max_size()];
+                self.sprite_layers.insert(
+                    z,
+                    Some(SpriteLayer {
+                        inner: LayerKindInner::Dense(DenseLayer::new(tiles)),
+                        entity: None,
+                    }),
+                )
+            }
+            LayerKind::Sparse => self.sprite_layers.insert(
+                z,
+                Some(SpriteLayer {
+                    inner: LayerKindInner::Sparse(SparseLayer::new(HashMap::default())),
+                    entity: None,
+                }),
+            ),
+        }
+    }
+
+    pub(crate) fn move_layer(&mut self, from_z: usize, to_z: usize) {
+        if self.sprite_layers[to_z].is_some() {
+            panic!("Sprite layer {} unexpectedly exists.", to_z);
+        }
+
+        self.sprite_layers[to_z] = self.sprite_layers[from_z].clone();
+        self.sprite_layers[from_z] = None;
+    }
+
+    pub(crate) fn remove_layer(&mut self, z_layer: usize) {
+        self.sprite_layers[z_layer] = None;
+    }
+
+    pub(crate) fn set_mesh(&mut self, z_layer: usize, mesh: Handle<Mesh>) {
+        let layer = self.sprite_layers[z_layer].as_mut().unwrap();
+        layer.inner.as_mut().set_mesh(mesh);
```

Comment:
The meshes can be 3d, in which case the entire chunk could use a single mesh. That would reduce the number of draw calls necessary to render a map, which is the first thing you always try to do to improve rendering performance. It would make the meshes larger so it might limit the maximum size of a chunk. I think this might be a good optimization to experiment with later.

Reply:
Hmm. Good idea. We should chat about that.